008-strimzi-ui.md

```diff
@@ -0,0 +1,119 @@
+
+```
+Scenario: Seeing all available topics:
+
+Given I am a user of the UI
+When I navigate to the topic listing
+Then see all topics in my Kafka cluster
+
+Scenario: Viewing a specific topic (when it exists in the cluster):
+Given I am a user of the UI
+When I navigate to the topic listing
+And I filter for topic 'SampleTopic'
+Then I see topic 'SampleTopic' in the topic list
+
+Scenario: Viewing a specific topic (when it does not exist in the cluster):
+Given I am a user of the UI
+When I navigate to the topic listing
+And I filter for topic 'SampleTopic'
+Then I am told topic 'SampleTopic' does not exist in the cluster
+```
+
+#### Follow on capabilities
+
+Additional Kafka capabilities which could be added later include (but would not be limited to):
+
+- Create, update and delete topics on the topics list page (recontextualised as a Topics page - as per mock up)
+- View Consumer group status for a given topic, or the whole cluster
+- View the Brokers in a given cluster, their configuration, and where appropriate, allow modification of broker configuration
+- Provide details of Kafka listener/bootstrap addresses, along with sample configuration, to allow streamline client connectivity
+
+If desired, additional Strimzi provided capabilities, such as Kafka Users, could be leveraged/added/managed through this UI also (for example, allowing access to the UI and its capabilities, general CRUD updates to them etc). 
+
+Finally, one capability which may be of interest (which may have bearing on how this is surfaced in Strimzi itself) is having a single UI, which can manage multiple Kafka clusters deployed via the Strimzi operator.
+
+In all of these cases, capabilities can be added in a prioritised order, and should be added in a phased manner themselves (for example, add a view of the brokers and their configuration, then the ability to modify select configuration).
+
+## Affected projects
+
+I would expect that the main development effort for a UI will be in a new repository in the Strimzi Github organisation. However, given the discussion in https://github.com/strimzi/strimzi-kafka-operator/issues/2540 , one option to source backend Kafka data would be via the [Bridge](https://github.com/strimzi/strimzi-kafka-bridge). As required, this could be extended to support the various Kafka backend calls needed to surface any required information in a UI. In addition, depending on how and where the UI is provided in Strimzi, changes may be needed in the [strimzi-kafka-operator](https://github.com/strimzi/strimzi-kafka-operator) to either define a UI CRD, or deploy a UI deployment if configured to do so via the operator. I would very much welcome discussion around this, and what would make the most sense for the Strimzi project.
```

Comment:
I would say that at this point the backend will be developed via a new admin server as per this PR https://github.com/strimzi/strimzi-admin/pull/4 even because providing GraphQL API is not part of the bridge roadmap which is more about adding different protocol heads (HTTP, AMQP, ..) to the Kafka protocol.

Reply:
One thought we had was to add new HTTP endpoints to bridge if/when needed, then the UI backend server provides a GraphQL layer that fetches from bridge.

If however the admin api proposed provides GraphQL, then the UI bridge endpoints can instead move to the admin api (and either go via kafka admin client, or via bridge).

Reply:
Yes, that could be a possibility as well. In the end, with my pretty low knowledge about GraphQL, the GQL server can connect to any kind of REST API service using different endpoints for fathering information with one model response to the GraphQL client.
The concern could be that at this point the bridge becomes a mandatory component for the clusters where people want the UI and that at same time it's used for both consumers/producers + UI administrative operations.
I was wondering if the bridge Admin API endpoint is more for administrative operations made by HTTP Kafka clients and not an admin UI. It's a topic for discussion. @strimzi/maintainers wdyt?